crates/weaver_common/src/result.rs

```diff
@@ -0,0 +1,180 @@
+    /// The operation failed with a fatal error. By definition, we can only have
+    /// one fatal error.
+    FatalErr(E),
+}
+
+impl<T, E> WResult<T, E>
+where
+    E: WeaverError<E> + Error + Diagnostic + Serialize + Send + Sync + 'static,
+{
+    /// Converts a [`WResult`] into a standard [`Result`], optionally capturing non-fatal errors.
+    pub fn capture_non_fatal_errors(
+        self,
+        non_fatal_errors: &mut Vec<DiagnosticMessage>,
+    ) -> Result<T, E> {
+        match self {
+            WResult::Ok(result) => Ok(result),
+            WResult::OkWithNFEs(result, nfes) => {
+                for non_fatal_error in nfes {
+                    non_fatal_errors.push(DiagnosticMessage::new(non_fatal_error));
+                }
+                Ok(result)
+            }
+            WResult::FatalErr(fatal_err) => Err(fatal_err),
+        }
+    }
+
+    /// Capture the warnings into the provided vector and return a [`WResult`]
+    /// without the warnings.
+    pub fn capture_warnings(self, diag_msgs: &mut DiagnosticMessages) -> WResult<T, E> {
+        if let WResult::OkWithNFEs(result, nfes) = self {
+            let (warnings, errors): (Vec<_>, Vec<_>) = nfes
+                .into_iter()
+                .partition(|e| matches!(e.severity(), Some(miette::Severity::Warning)));
+            let warnings: Vec<_> = warnings.into_iter().map(DiagnosticMessage::new).collect();
+            diag_msgs.extend_from_vec(warnings);
+            if errors.is_empty() {
+                WResult::Ok(result)
+            } else {
+                WResult::OkWithNFEs(result, errors)
+            }
+        } else {
+            self
+        }
+    }
+
+    /// Return a [`WResult`] without the warnings.
+    pub fn ignore_warnings(self) -> WResult<T, E> {
+        match self {
+            WResult::OkWithNFEs(result, non_fatal_errors) => {
+                // Remove warnings from the non-fatal errors.
+                let errors: Vec<_> = non_fatal_errors
+                    .into_iter()
+                    .filter(|e| !matches!(e.severity(), Some(miette::Severity::Warning)))
+                    .collect();
+                if errors.is_empty() {
+                    WResult::Ok(result)
+                } else {
+                    WResult::OkWithNFEs(result, errors)
+                }
+            }
+            _ => self,
+        }
+    }
+
+    /// Calls a function with a reference to the contained value if [`Ok`].
+    ///
+    /// Returns the original result.
+    pub fn inspect<F: FnOnce(&T, Option<&[E]>)>(self, f: F) -> Self {
+        match &self {
+            WResult::Ok(result) => f(result, None),
+            WResult::OkWithNFEs(result, nfes) => f(result, Some(nfes)),
+            WResult::FatalErr(_) => {}
+        }
+
+        self
+    }
+
+    /// Converts a [`WResult`] into a standard [`Result`], potentially
+    /// aggregating non-fatal errors into a single error.
+    pub fn into_result(self) -> Result<T, E> {
```

Comment:
NIt: I actually think ignoring non_fatal_errors could be the default here.  I'm not certain, but I'd suggest `into_result_*` for all methods instead of an `into_result` that isn't explicit:

- `into_result_with_non_fatal` -> `Result<(T,Vec<E>), E>`
- `into_result_ignoring_non_fatal`  -> `Result<T,E>`
- `into_result_failing_non_fatal` -> `Result<T,E>`

Or some such.

Reply:
I agree. It’s wordy, but at least it avoids ambiguities in use.